# 1.9 Strings

Gezien we veel met basistypes zullen werken en dus ook met strings, krijgen ze hieronder extra aandacht. Strings zijn eigenlijk al net iets meer dan basistypes: we kunnen ze echt gebruiken als objecten (uiteraard zijn alle variabelen en elementen in Python objecten) met bijhorende methodes. Verder is een string, net als een set/lijst/..., een itereerbaar object. Meer nog, net als een lijst hebben strings dezelfde indexering en kunnen we vierkante haken gebruiken om hen uit elkaar te halen per index.

Verder zijn strings ook belangrijk omdat we uiteindelijk het meeste van onze output nog altijd printen naar de console. Daarom is het belangrijk te snappen hoe tekst verwerkt wordt in Python. Hieronder worden zowel de indexering, als het gebruik van de `+`-operator geïllustreerd, alsook enkele handige functies. Zo kan de `len()` functie bij strings worden gebruikt om het aantal elementen te leren en `in` om te controleren of substrings in de string voorkomen:

In [1]:
a = 'Softwareontwerp'
b = 'en programmeren'
print(a + b + b) # concatenatie
print(a + b + '_' + str(5)) # conversie van een integer
print(a[2:7:2]) # indexering
print(a[::-1]) # omgekeerd
print(len(a)) # lengte van de string
print('oftware' in a)
print('oftware' in b)

Softwareontwerpen programmerenen programmeren
Softwareontwerpen programmeren_5
fwr
prewtnoerawtfoS
15
True
False


## Methodes

Er zijn heel wat methodes die kunnen gebruikt worden bij strings. Je gebruikt ze door een punt (`.`) achter een variabele van het type string te zetten, gevolgd door de naam van de methode en de haakjes voor de argumenten, net als bij functies. Deze zullen we vaak gebruiken:
- `capitalize()` maakt alles lower case behalve de eerste letter
- `lower()` maakt alle letters lower case
- `upper()` maakt alle letters upper case
- `title()` geeft aan alle aparte woorden een hoofdletter aan het begin
- `count(arg1)` kan, gegeven een substring als argument `arg1`, het aantal voorkomens van die substring tellen
- `strip()` haalt alle whitespace (dus: spaties, tabs en nieuwe lijn-symbolen zoals '\n') weg uit een string aan de uiteinden
- `lstrip()`/`rstrip()` doen hetzelfde als `strip()` maar enkel aan de linker- respectievelijk rechterkant van de string

In [2]:
a = 'programmeren Is Leuk'
b = '  programmeren is leuk als je het veel probeert     '
print(a.capitalize())
print(a.lower())
print(a.upper())
print(a.title())
print(b.count('ee'))
print('--', b.strip(), '--')
print('--', b.lstrip(), '--')
print('--', b.rstrip(), '--')

Programmeren is leuk
programmeren is leuk
PROGRAMMEREN IS LEUK
Programmeren Is Leuk
2
-- programmeren is leuk als je het veel probeert --
-- programmeren is leuk als je het veel probeert      --
--   programmeren is leuk als je het veel probeert --


Verder zijn er nog andere ingebouwde methodes die handig kunnen zijn:
- `find(arg1)` geeft de index van het eerste voorkomen van de meegegeven substring `arg1` terug en geeft -1 indien niets gevonden is
- `replace(arg1, arg2)` vervangt elk voorkomen van een substring (`arg1`) door een andere string (`arg2`)
- `rfind(arg1)` vind het eerste voorkomen van een substring `arg1` van rechts te beginnen (de teruggegeven index is de startpositie van de substring) en geeft -1 indien niets gevonden is
- `split(arg1)` splits een string gegeven een delimiter/scheidingsteken `arg1`, meegegeven als argument, en geeft een lijst van substrings terug die door het teken gescheiden werden (veel gebruikt: splits een lange string met komma's in een lijst van aparte woorden)
- `rsplit()` splitst een string op white spaces

Daarnaast zijn er ook nog 2 functies die het toelaten de inhoud van de string te checken:
- `isalnum()` geeft een boolean terug die zegt of alle tekens in de string alfabetisch of numeriek zijn, handig om bv. extra vreemde tekens zoals '@' te vinden
- `isalpha()` geeft een boolean terug die zegt of alle tekens alfabetisch zijn, handig dus om te zien of er getallen in de string staan

In [3]:
a = '123 abc defgg gg'
print(a.find('gg'))
print(a.replace('gg', 'ffff'))
print(a.rfind('gg'))
print(a.split('abc'))
print(a.rsplit())
b = '123abc'
c = 'abc'
d = '@bc12'
print(b.isalnum())
print(b.isalpha())
print(c.isalnum())
print(c.isalpha())
print(d.isalnum())
print(d.isalpha())

11
123 abc defffff ffff
14
['123 ', ' defgg gg']
['123', 'abc', 'defgg', 'gg']
True
False
True
True
False
False


## Printen

De `print()` functie helpt ons om te bevatten wat er in onze variabelen zit. We kunnen ze gebruiken zoals voorheen, door strings mee te geven, maar we kunnen ook meerdere strings meegeven als aparte argumenten gescheiden door een komma. Dat maakt dat we zelfs bv. integers niet moeten omzetten naar strings, dit gebeurt automatisch. Standaard zal de `print()` functie met komma's tussen de argumenten alles scheiden met een spatie, maar die scheiding kan aangepast worden door een expliciet `sep` argument toe te voegen (zie de latere notebooks rond argumenten om te snappen wat argumenten met een specifieke naam zijn). Ook kunnen we variabelen rechtstreeks in onze strings gebruiken door er accolades rond te zetten, zolang we eerst een 'f' voor de string zetten. Die geeft Python het signaal dat er variabelen kunnen voorkomen tussen accolades. 

Ten slotte kunnen ook bepaalde placeholders gebruikt worden die getallen kunnen formatteren. Dit gebeurt met een `%`-symbool. In de string duidt deze aan dat er een getal volgt. We gebruiken '%d' om een geheel getal aan te duiden.
Bij een kommagetal gebruiken we een '.Xf' om aan te duiden hoeveel X getallen er na de komma van een `float` (f) moeten getoond worden. Na de string wordt dan via het %-symbool tussen haakjes het getal/de getallen meegegeven.

In [4]:
a = 2
b = 'Johannes'
print('Zoals voorheen gezien, maar nu met 2 extra argumenten', a, b)
print('Met een andere', 'separator ertussen', sep='---___---')
print('Concatenatie met +-operator vereist juiste types ' + str(a))
print(f'{b} geeft {a} vakken')
print('Ik kocht %d bananen in de winkel aan %.1f euro' % (2, 0.35)) # merk op dat 0.35 wordt afgerond

Zoals voorheen gezien, maar nu met 2 extra argumenten 2 Johannes
Met een andere---___---separator ertussen
Concatenatie met +-operator vereist juiste types 2
Johannes geeft 2 vakken
Ik kocht 2 bananen in de winkel aan 0.3 euro


Formattering kan nog iets verder gaan. I.p.v. gebruik te maken van %-symbolen kunnen we ook de `format()` methode gebruiken. In de string zelf gebruiken we accolades met nummers om aan te duiden welke elementen uit `format()` waar moeten komen. We kunnen zelfs lijsten in `format()` gebruiken en de verschillende elementen op verschillende plaatsen in een string opvragen. Merk op dat het aantal argumenten van `format()` niet vastligt. Daarnaast kunnen nog altijd dezelfde formatteringscodes gebruikt worden zoals `0:.2f` hieronder, dat het eerste (op index 0) element uit `format()` omzet naar een kommagetal met 2 cijfers na de komma:

In [5]:
a = 100
b = 0.222
c = 'aandelen'
print('Ik kocht {0} {1} aan {2}'.format(a, c, b))
print('Ik kocht {2} {1} aan {0}'.format(a, c, b))

# merk op hoe hier een lijst het eerste argument vormt en zo ook wordt opgeroepen onder index 0 als 0[0] en 0[1]
# de twee elementen van die lijst op index 0
# 0.12 is het tweede argument (op index 1 dus)
print('Ik kocht {0[0]} {0[1]} aan {1}'.format([100, 'aandelen'], 0.12))
print('Ik kocht een aandeel aan {0:.2f}'.format(10200.2211))

Ik kocht 100 aandelen aan 0.222
Ik kocht 0.222 aandelen aan 100
Ik kocht 100 aandelen aan 0.12
Ik kocht een aandeel aan 10200.22


Het aantal spaties dan gebruikt wordt om getallen te printen kan ook gecontroleerd worden door het gebruik van een extra getal (maar dan voor de komma bij floats). Hieronder worden er bijvoorbeeld 40 white spaces gebruikt om een geheel getal te printen (het getal 10) en daarna 20 om het kommagetal 10200.22 te printen. Dit is handig als je getallen van verschillende lengtes wilt aligneren, zoals in de for-loop waar uiteindelijk een getal boven de 10 ook mooi rechts uitgelijnd is met de rest:

In [6]:
print('Ik kocht %40d aandelen' % (10))
print('Ik kocht een aandeel aan {0:20.2f}'.format(10200.2211))
for i in range(0, 11):
    print('i is {0:2}'.format(i))

Ik kocht                                       10 aandelen
Ik kocht een aandeel aan             10200.22
i is  0
i is  1
i is  2
i is  3
i is  4
i is  5
i is  6
i is  7
i is  8
i is  9
i is 10


## Input 

Naast uitprinten kunnen we ook dingen invoeren via de console. Dit gebeurt telkens door strings in te voeren en die eventueel om te zetten indien nodig (bijvoorbeeld naar getallen).
We gebruiken hiervoor de `input()`-functie. Die kan ofwel zonder argument gebruikt worden, waarbij je input levert in een leeg veld, of door een argument mee te geven dat omschrijft wat in te voeren:

In [10]:
ingelezen_string = input()
print('---', ingelezen_string, '---')

ingelezen_string = input('Geef een nummer ')
print('Omgezet en gedeeld door 3:', int(float(ingelezen_string)/3))

ik geef een string in
--- ik geef een string in ---
Geef een nummer 12
Omgezet en gedeeld door 3: 4


## Groter voorbeeld

Hieronder worden wat concepten bij elkaar gebracht in een groter voorbeeld:

In [2]:
zin = "Vanavond eten we spaghetti"
print(zin)

letters = set(zin) # set verwijdert de duplicaten en houdt enkel de unieke letters over
teller = 0
output_string = ''
for teken in letters:
    print('\nTeken:', teken)
    print('Teller:', teller)
    print('Count:', zin.count(teken))
    print('(1)', output_string)
    output_string += "{0:3s} komt {1:3d} keer voor \t".format(teken, zin.count(teken))
    teller += 1
    print('(2)', output_string)
    if teller % 3 == 0: # om de 4 tekens gaan we de huidige output printen en een nieuwe lijn beginnen
        print('Teller deelbaar door 3, nieuwe lijn')
        print(output_string)
        output_string = ''

Vanavond eten we spaghetti

Teken: s
Teller: 0
Count: 1
(1) 
(2) s   komt   1 keer voor 	

Teken: w
Teller: 1
Count: 1
(1) s   komt   1 keer voor 	
(2) s   komt   1 keer voor 	w   komt   1 keer voor 	

Teken: i
Teller: 2
Count: 1
(1) s   komt   1 keer voor 	w   komt   1 keer voor 	
(2) s   komt   1 keer voor 	w   komt   1 keer voor 	i   komt   1 keer voor 	
Teller deelbaar door 3, nieuwe lijn
s   komt   1 keer voor 	w   komt   1 keer voor 	i   komt   1 keer voor 	

Teken: e
Teller: 3
Count: 4
(1) 
(2) e   komt   4 keer voor 	

Teken: a
Teller: 4
Count: 3
(1) e   komt   4 keer voor 	
(2) e   komt   4 keer voor 	a   komt   3 keer voor 	

Teken: p
Teller: 5
Count: 1
(1) e   komt   4 keer voor 	a   komt   3 keer voor 	
(2) e   komt   4 keer voor 	a   komt   3 keer voor 	p   komt   1 keer voor 	
Teller deelbaar door 3, nieuwe lijn
e   komt   4 keer voor 	a   komt   3 keer voor 	p   komt   1 keer voor 	

Teken: t
Teller: 6
Count: 3
(1) 
(2) t   komt   3 keer voor 	

Teken: g
Teller: 7
Count: